Add explicit to_int cast to ^Int SMT hook

Baltoli opened the pull request:
This was recommended as a workaround for #2154 and #1957 by the Z3
maintainers (https://github.com/Z3Prover/z3/issues/5243).

The original issue arises because Z3 cannot assume that the exponent is
non-negative, and so must infer that the result could potentially have
type Real after evaluating the exponentiation. By adding the to_int
cast, we opt in to integer semantics everywhere ^Int is used.

Additionally, this commit adds a simplification rule for integer
exponentiation that sends negative exponents to zero directly from K.

cos commented:
you can update the outputs with `--update-out`

msaxena2 commented:
@cos @ehildenb @lucaspena Please review. Also, the constraints are now only printed when search is called with the --print-constraints flag.

ehildenb commented on java-backend/src/main/java/org/kframework/backend/java/symbolic/InitializeRewriter.java:
```diff
@@ -33,10 +36,12 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
```
@msaxena2 It builds fine on my system without the extra import statements.

cos commented on java-backend/src/main/java/org/kframework/backend/java/symbolic/InitializeRewriter.java:
```diff
@@ -129,33 +134,30 @@ public RewriterResult execute(K k, Optional<Integer> depth) {
 
         @Override
-        public List<? extends Map<? extends KVariable,? extends K>> match(K k, org.kframework.definition.Rule rule) {
+        public List<Tuple2<? extends Map<? extends KVariable, ? extends K>, ? extends K>> match(K k, org.kframework.definition.Rule rule) {
```
it may not be clear what the second projection of the tuple is. a new class holding the map and the remaining constraints, or some javadoc to this method would help.

cos commented on java-backend/src/main/java/org/kframework/backend/java/symbolic/InitializeRewriter.java:
```diff
@@ -129,33 +134,30 @@ public RewriterResult execute(K k, Optional<Integer> depth) {
 
         @Override
-        public List<? extends Map<? extends KVariable, ? extends K>> search(K initialConfiguration, Optional<Integer> depth, Optional<Integer> bound, Rule pattern, SearchType searchType) {
+        public List<Tuple2<? extends Map<? extends KVariable, ? extends K>, ? extends K>> search(K initialConfiguration, Optional<Integer> depth, Optional<Integer> bound, Rule pattern, SearchType searchType) {
```
same, class or javadoc

cos commented on java-backend/src/main/java/org/kframework/backend/java/symbolic/InitializeRewriter.java:
```diff
@@ -129,33 +134,30 @@ public RewriterResult execute(K k, Optional<Integer> depth) {
 
 
-        public Tuple2<RewriterResult, List<? extends Map<? extends KVariable, ? extends K>>> executeAndMatch(K k, Optional<Integer> depth, Rule rule) {
+        public Tuple2<RewriterResult, List<Tuple2<? extends Map<? extends KVariable, ? extends K>, ? extends K>>> executeAndMatch(K k, Optional<Integer> depth, Rule rule) {
```
same, class or javadoc.

cos commented on java-backend/src/main/java/org/kframework/backend/java/symbolic/SymbolicRewriter.java:
```diff
@@ -402,7 +401,7 @@ public static ConstrainedTerm buildResult(
      */
     private boolean addSearchResult(
-            HashSet<Substitution<Variable, Term>> searchResults,
+            HashSet<Tuple2<Substitution<Variable, Term>, K>> searchResults,
```
same as above, class or javadoc

cos commented on java-backend/src/main/java/org/kframework/backend/java/symbolic/SymbolicRewriter.java:
```diff
@@ -430,7 +433,7 @@ private boolean addSearchResult(
      * @return a list of substitution mappings for results that matched the pattern
      */
-    public Set<Substitution<Variable, Term>> search(
+    public Set<Tuple2<Substitution<Variable, Term>, K>> search(
```
same as above, class or javadoc

cos replied:
also, the javadoc seems to be obsolete -- might help to update it anyhow

cos commented on java-backend/src/main/java/org/kframework/backend/java/symbolic/SymbolicRewriter.java:
```diff
@@ -439,7 +442,7 @@ private boolean addSearchResult(
         stopwatch.start();
 
-        HashSet<Substitution<Variable, Term>> searchResults = Sets.newHashSet();
+        HashSet<Tuple2<Substitution<Variable, Term>, K>> searchResults = Sets.newHashSet();
```
same

cos commented on kernel/src/main/java/org/kframework/debugger/DebuggerMatchResult.java:
```diff
@@ -13,12 +14,12 @@
  */
 public class DebuggerMatchResult {
-    private final List<? extends Map<? extends KVariable, ? extends K>> substitutions;
+    private final List<Tuple2<? extends Map<? extends KVariable, ? extends K>, ? extends K>>  substitutions;
```
same -- it looks like a separate class would be the best choice. as we want to work with K conjunctions in the future, it may be best to use that already, and document the methods.

cos commented on kernel/src/main/java/org/kframework/krun/KRun.java:
```diff
@@ -125,27 +126,39 @@ public int run(CompiledDefinition compiledDef, KRunOptions options, Function<Mod
-                .collect(Collectors.toSet());
+        Set<Map<? extends KVariable, ? extends K>> searchResult = new HashSet<>();
+        ArrayList<K> constraints = new ArrayList<>();
+        result.getSearchList().forEach(tuple -> {
```
"tuple" is not informative

cos commented on kernel/src/main/java/org/kframework/krun/KRun.java:
```diff
@@ -125,27 +126,39 @@ public int run(CompiledDefinition compiledDef, KRunOptions options, Function<Mod
-        int i = 1;
-        List<String> results = new ArrayList<>();
+        int i = 0;
+        List<Tuple2<String, String>> results = new ArrayList<>();
```
what's the first and the second string?

cos commented on kernel/src/main/java/org/kframework/krun/KRunOptions.java:
```diff
@@ -160,6 +160,9 @@ public OutputModeConverter(String optionName) {
     public boolean search = false;
 
+    @Parameter(names={"--print-constraint", "--print-constraints"}, description="If specified with search, prints the constraints for each solution.")
+    public boolean printConstraint = false;
```
if there is no other problem, let's always print the constraints (unless they are simply true)

cos commented on kore/src/main/scala/org/kframework/rewriter/Rewriter.scala:
```diff
@@ -23,7 +23,7 @@ trait Rewriter {
 
 
-  def `match`(k: kore.K, rule: Rule): java.util.List[_ <: java.util.Map[_ <: kore.KVariable, _ <: kore.K]]
+  def `match`(k: kore.K, rule: Rule): java.util.List[Tuple2[_ <: java.util.Map[_ <: kore.KVariable, _ <: kore.K], _ <: kore.K]]
```
again, considering that this is the K API, we really should go with the conjunction. I know we discussed this before, but I didn't realize that the tuple will be leaking all the way to the API.

cos replied:
Yes. As @andreistefanescu said below, you want a disjunction of conjunctions. Each map becomes a conjunction, and the list becomes a disjunction.

cos commented on kore/src/main/scala/org/kframework/rewriter/Rewriter.scala:
```diff
@@ -34,9 +34,9 @@ trait Rewriter {
    * @return A list of substitutions, denoting all the configurations matching the given rule.
    */
-  def search(initialConfiguration: kore.K, depth: Optional[Integer], bound: Optional[Integer], pattern: Rule, searchType: SearchType): java.util.List[_ <: java.util.Map[_ <: kore.KVariable, _ <: kore.K]]
+  def search(initialConfiguration: kore.K, depth: Optional[Integer], bound: Optional[Integer], pattern: Rule, searchType: SearchType): java.util.List[Tuple2[_ <: java.util.Map[_ <: kore.KVariable, _ <: kore.K], _ <: kore.K]]
```
same

cos commented on kore/src/main/scala/org/kframework/rewriter/Rewriter.scala:
```diff
@@ -34,9 +34,9 @@ trait Rewriter {
+  def search(initialConfiguration: kore.K, depth: Optional[Integer], bound: Optional[Integer], pattern: Rule, searchType: SearchType): java.util.List[Tuple2[_ <: java.util.Map[_ <: kore.KVariable, _ <: kore.K], _ <: kore.K]]
 
-  def executeAndMatch(k: kore.K, depth: Optional[Integer], rule: Rule): Tuple2[RewriterResult, java.util.List[_ <: java.util.Map[_ <: kore.KVariable, _ <: kore.K]]]
+  def executeAndMatch(k: kore.K, depth: Optional[Integer], rule: Rule): Tuple2[RewriterResult, java.util.List[Tuple2[_ <: java.util.Map[_ <: kore.KVariable, _ <: kore.K], _ <: kore.K]]]
```
same

andreistefanescu commented:
The list of conjunctions can be represented by a disjunction.

On Mon, Oct 10, 2016, 5:08 PM Manasvi Saxena <notifications@github.com>
wrote:

> *@msaxena2* commented on this pull request.
> ------------------------------
>
> In kore/src/main/scala/org/kframework/rewriter/Rewriter.scala
> <https://github.com/kframework/k/pull/2210>:
>
> > @@ -23,7 +23,7 @@ trait Rewriter {
>    def execute(k: kore.K, depth: Optional[Integer]): RewriterResult
>
>
> -  def `match`(k: kore.K, rule: Rule): java.util.List[_ <: java.util.Map[_ <: kore.KVariable, _ <: kore.K]]
> +  def `match`(k: kore.K, rule: Rule): java.util.List[Tuple2[_ <: java.util.Map[_ <: kore.KVariable, _ <: kore.K], _ <: kore.K]]
>
> I agree that a conjunction should be the way to go. The rewriter returns a
> list of maps, where each map represents a substitution. I'm not sure how we
> can go from a list of maps to a conjunction, unless we make the rewriter
> itself return the map as a K term (which can then be used in a conjunction
> with the constraint). Is that what you had in mind as well?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kframework/k/pull/2210>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE3jMcWf806yDZeU58AWvMD09rXnOnAHks5qytNugaJpZM4KNFzP>
> .
>

msaxena2 commented:
@cos Please have a look.

cos commented on kernel/src/main/java/org/kframework/krun/KRun.java:
```diff
@@ -157,23 +211,29 @@ private void printSearchResult(SearchResult result, KRunOptions options, Compile
-    public static int getExitCode(KExceptionManager kem, List<? extends Map<? extends KVariable, ? extends K>> res) {
-        if (res.size() != 1) {
-            kem.registerCriticalWarning("Found " + res.size() + " solutions to exit code pattern. Returning 112.");
+    public static int getExitCode(KExceptionManager kem, K res) {
```
what is this "exitCode"?

msaxena2 replied:
Already existing code, seems like it matches some pre-existing pattern on the final configuration. Should probably get rid of it in a separate PR.

cos commented on kernel/src/main/java/org/kframework/krun/KRun.java:
```diff
@@ -124,30 +123,85 @@ public int run(CompiledDefinition compiledDef, KRunOptions options, Function<Mod
+    private StringBuilder filterAnonVarsAndPrint(K result, Set<String> filterSet, CompiledDefinition compiledDef, KRunOptions options) {
+        StringBuilder sb = new StringBuilder();
+        ByteArrayOutputStream bs = new ByteArrayOutputStream();
+        if (result instanceof KApply && ((KApply) result).klabel().toString().equals(KLabels.AND)) {
```
do not check string-based equality

cos commented on java-backend/src/main/java/org/kframework/backend/java/symbolic/SymbolicRewriter.java:
```diff
@@ -525,15 +552,41 @@ private boolean addSearchResult(
-        }).collect(Collectors.toSet());
+    private void flattenList(List<K> unflat, List<K> flat) {
+        unflat.forEach(x -> {
+            if (x instanceof KItem && ((KItem) x).klabel().toString().equals(KLabels.AND)) {
```
do not check equality based on `toString`

cos commented on java-backend/src/main/java/org/kframework/backend/java/symbolic/SymbolicRewriter.java:
```diff
@@ -525,15 +552,41 @@ private boolean addSearchResult(
+    }
+
+    private K processConjuncts(ConjunctiveFormula conjunct) {
+        if (conjunct.klabel().toString().equals(KLabels.AND)) {
```
same, `toString`

cos commented on kernel/src/main/java/org/kframework/krun/KRun.java:
```diff
@@ -124,30 +123,85 @@ public int run(CompiledDefinition compiledDef, KRunOptions options, Function<Mod
+                if (x.getClass().toString().contains("BoolToken")) {
+                    return false;
+                }
+                if ((options.pattern != null) && x instanceof KApply && ((KApply) x).klabel().toString().equals(KLabels.EQUALS)) {
```
`toString`

cos commented on kernel/src/main/java/org/kframework/krun/KRun.java:
```diff
@@ -124,30 +123,85 @@ public int run(CompiledDefinition compiledDef, KRunOptions options, Function<Mod
+        if (result instanceof KApply && ((KApply) result).klabel().toString().equals(KLabels.AND)) {
+            List<K> conjunctions = mutable(Assoc.flatten(KLabel(KLabels.AND), immutable(((KApply) result).items()), KLabel(KLabels.ML_TRUE)));
+            conjunctions = conjunctions.stream().filter(x -> {
+                if (x.getClass().toString().contains("BoolToken")) {
```
don't use a string for the check

cos commented on kernel/src/main/java/org/kframework/krun/KRun.java:
```diff
@@ -124,30 +123,85 @@ public int run(CompiledDefinition compiledDef, KRunOptions options, Function<Mod
-            prettyPrintSubstitution(substitution, result.getParsedRule(), compiledDef, options.output, v -> sb.write(v, 0, v.length));
-            //Note that this is actually unsafe, but we are here assuming that --search is not used with --output binary
-            results.add(new String(sb.toByteArray()));
+    private StringBuilder filterAnonVarsAndPrint(K result, Set<String> filterSet, CompiledDefinition compiledDef, KRunOptions options) {
```
if possible, separate the filter from the print

cos commented on kernel/src/main/java/org/kframework/krun/KRun.java:
```diff
@@ -157,23 +211,29 @@ private void printSearchResult(SearchResult result, KRunOptions options, Compile
-                    vars.add(Integer.valueOf(((KToken) t).s()));
-                } catch (NumberFormatException e) {
-                    throw KEMException.criticalError("Exit code found was not in the range of an integer. Found: " + ((KToken) t).s(), e);
+        new VisitK() {
```
extract visitor to someplace else

msaxena2 replied:
Let's let this be as is for now. We'll probably get rid of this function in a later PR.

cos commented on kernel/src/main/java/org/kframework/krun/KRun.java:
```diff
@@ -124,30 +123,85 @@ public int run(CompiledDefinition compiledDef, KRunOptions options, Function<Mod
-            sb.append(solution);
-            sb.append("\n");
+        if (options.pattern != null) {
+            new VisitK() {
```
separate visitor -- this method is kind of large

cos commented on kore/src/main/scala/org/kframework/rewriter/Rewriter.scala:
```diff
@@ -34,9 +34,9 @@ trait Rewriter {
    * @return A list of substitutions, denoting all the configurations matching the given rule.
    */
-  def search(initialConfiguration: kore.K, depth: Optional[Integer], bound: Optional[Integer], pattern: Rule, searchType: SearchType): java.util.List[_ <: java.util.Map[_ <: kore.KVariable, _ <: kore.K]]
+  def search(initialConfiguration: kore.K, depth: Optional[Integer], bound: Optional[Integer], pattern: Rule, searchType: SearchType, subsitution:Boolean): kore.K
```
I really don't like how this method behaves differently based on the substitution flag. Let's leave it like this for now so we can merge, but please add an issue and we should fix this with the very first opportunity.

msaxena2 commented:
@cos Please re-review.

Baltoli committed:
Add explicit to_int cast to ^Int SMT hook

This was recommended as a workaround for #2154 and #1957 by the Z3
maintainers (https://github.com/Z3Prover/z3/issues/5243).

The original issue arises because Z3 cannot assume that the exponent is
non-negative, and so must infer that the result could potentially have
type Real after evaluating the exponentiation. By adding the to_int
cast, we opt in to integer semantics everywhere ^Int is used.

Additionally, this commit adds a simplification rule for integer
exponentiation that sends negative exponents to zero directly from K.

Baltoli committed:
Fix unused variable

radumereuta commented:
Can you do some tests with different Z3 versions? We want to know which version to recommend in the future.
In my original issue, I found that z3-4.8.10 wasn't working, but after that, it might work.

Baltoli commented:
Yes, I'll verify that this works with different releases of Z3.

ehildenb commented on k-distribution/include/kframework/builtin/domains.md:
```diff
@@ -930,7 +930,7 @@ You can:
   syntax Int ::= "~Int" Int                     [function, klabel(~Int_), symbol, functional, latex(\mathop{\sim_{\scriptstyle\it Int}}{#1}), hook(INT.not), smtlib(notInt)]
                > left:
-                 Int "^Int" Int                 [function, klabel(_^Int_), symbol, left, smt-hook(^), latex({#1}\mathrel{{\char`\^}_{\!\scriptstyle\it Int}}{#2}), hook(INT.pow)]
+                 Int "^Int" Int                 [function, klabel(_^Int_), symbol, left, smt-hook((to_int (^ #1 #2))), latex({#1}\mathrel{{\char`\^}_{\!\scriptstyle\it Int}}{#2}), hook(INT.pow)]
```
We need to not use this encoding if the second argument is `<Int 0`. Because this is declared in a different module than the rule you added below, we are not guaranteed that a user will have the below simplification rule apply prior to sending to SMT.

If the second argument is `<Int 0`, then the overall result should be `0`. We could encode this to Z3 as `(ite (< #2 0) (0) (to_int (^ #1 #2)))`.

ehildenb replied:
We also need to check that `to_int` is _truncating_, not _rounding_.

Baltoli replied:
For positive arguments, the definition of `to_int` is truncating rather than rounding (see documentation [here](http://smtlib.cs.uiowa.edu/theories-Reals_Ints.shtml))

Baltoli committed:
Conditional application of SMT encoding

Baltoli committed:
Merge branch 'master' into power-smt-hook

Baltoli commented:
As Radu identified in the original issue, the `to_int` workaround requires a newer version of Z3 (`>= 4.8.11` in my testing - see table below, where I tested the branch against source builds of Z3). The CI failure for this change seems to be because Ubuntu 20.04 supplies Z3 4.8.7 by default.

This means that this change as proposed would require an upgrade of the required Z3 version, which may not be practical.

| Z3 Version | Working? |
| -------------: | :-------------: |
| `4.8.7` | ✗ |
| `4.8.8` | ✗ |
| `4.8.9` | ✓ |
| `4.8.10` | ✗ |
| `4.8.11` | ✓ |
| `4.8.12` | ✓ |
| `HEAD` | ✓ |

radumereuta commented:
That's a problem because we need to decide which versions of Z3 we want to support.
And we need to synchronize with the haskell backend.
@dwightguth @ehildenb any thoughts? Or should we have this discussion in the front-end meeting?

ehildenb commented:
Hmmm, I guess we should test that all the existing semantics work with the bumped version, before going forward with it.

This will become a hard requirements for users then, because if they want to use the Haskell backend and the `^Int` symbol they'll build smt queries with that symbol present.

radumereuta commented:
Let's try to update to `z3-4.8.11`.
We need to update the docker files to the semantics to see if they all pass.
evm-semantics
wasm-semantics
Michelson-semantics
Algorand-sc-semantics
IELE-semantics
If the docker file doesn't specify the Z3 version, ping Everett.

Baltoli commented:
Superseded by #2225 via cherry-picked commits; closing
